File: zig/src/cpu.zig
const std = @import("std");
const Memory = @import("memory.zig").Memory;
const Display = @import("display.zig").Display;
const Keyboard = @import("keyboard.zig").Keyboard;
const Audio = @import("audio.zig").Audio;

const SPRITE_CHARS_ADDR = 0x0000;
var rnd = std.rand.DefaultPrng.init(0);

pub const CPU = struct {
    memory: Memory,
    delayTimer: u8,
    soundTimer: u8,
    microseconds: i32,
    pc: u16,
    registers: [16]u8,
    index: u16,
    sp: u8,
    stack: [16]u16,


    const Self = @This();

    pub fn init() Self {
        return CPU {
            .memory = Memory.init(),
            .delayTimer = 0,
            .soundTimer = 0,
            .microseconds = 0,
            .pc = 0x0200,
            .registers = [_]u8{0x00} ** 16,
            .index = 0x0000,
            .sp = 0,
            .stack = [_]u16{0x0000} ** 16,
        };
    }

    pub fn tick(self: *Self, display: *Display, keyboard: *Keyboard, audio: *Audio) void {
        if(self.delayTimer > 0){
            self.delayTimer -= 1;
        }
        if(self.soundTimer > 0){
            self.soundTimer -= 1;
            if(audio.isPlaying() == false){
                audio.start();
            }
        } else {
            if(audio.isPlaying()){
                audio.stop();
            }
        }

        self.microseconds += 16666;
        var delta: i32 = 0;
        while(self.microseconds > 0) : (self.microseconds -= delta){
            delta = self.emulateCycle(display, keyboard);
            if(delta == 0) {
                break;
            }
        }
    }

    pub fn emulateCycle(self: *Self, display: *Display, keyboard: *Keyboard) i32 {
        var opcode = self.getOpCode();
        return self.execute(opcode, display, keyboard);   
    }

    fn getOpCode(self: *Self) u16 {
        var opcode: u16 = @intCast(u16, self.memory.get(self.pc)) << 8 | self.memory.get(self.pc + 1);
        self.pc += 2;
        return opcode;
    }

    fn execute(self: *Self, opcode: u16, display: *Display, keyboard: *Keyboard) i32 {
        var x: u8 = @intCast(u8, opcode >> 8 & 0x0F);
        var y: u8 = @intCast(u8, opcode >> 4 & 0x0F);
        var n: u8 = @intCast(u8, opcode & 0x000F);
        var nn: u8 = @intCast(u8, opcode & 0x00FF);
        var nnn: u16 = opcode & 0x0FFF;
        switch(opcode & 0xF000) {
            0x0000 => switch (opcode & 0x00FF) {
                    0x00E0 => return opClearScreen(display),
                    0x000E => return self.opReturn(),
                    0x00EE => return self.opReturnFromSubroutine(),
                    else => return 0,
                },
            0x1000 => return self.opJump(nnn),
            0x2000 => return self.opJumpToSubroutine(nnn),
            0x3000 => return self.opSkipIfVxEqualsNn(x,nn),
            0x4000 => return self.opSkipIfVxNotEqualsNn(x, nn),
            0x5000 => return self.opSkipIfVxEqualsVy(x,y),
            0x6000 => return self.opSetRegisterXToNN(x, nn),
            0x7000 => return self.opAddNNToX(x, nn),
            0x8000 => switch (opcode & 0x000F) {
                    0x0000 => return self.opSetVxToValueOfVy(x, y),
                    0x0001 => return self.opBinaryOr(x, y),
                    0x0002 => return self.opBinaryAnd(x, y),
                    0x0003 => return self.opBinaryXor(x, y),
                    0x0004 => return self.opAddWithCarry(x, y),
                    0x0005 => return self.opSubtractYFromX(x, y),
                    0x0006 => return self.opShiftRight(x),
                    0x0007 => return self.opSubtractXFromY(x, y),
                    0x000E => return self.opShiftLeft(x),
                    else => return 0,
                },
            0x9000 => return self.opSkipIfVxNotEqualsVy(x, y),
            0xA000 => return self.opSetIndexToNNN(nnn),
            0xB000 => return self.opJumpWithOffset(nnn),
            0xC000 => return self.opRandom(x, nn),
            0xD000 => return self.opDisplay(x, y, n, display),
            0xE000 => switch (opcode & 0x00FF) {
                    0x009E => return self.opSkipIfKeyPressed(x, keyboard),
                    0x00A1 => return self.opSkipIfNotKeyPressed(x, keyboard),
                    else => return 0,
                },
            0xF000 => switch (opcode & 0x00FF) {
                    0x0007 => return self.opGetDelayTimer(x),
                    0x000A => return self.opGetKey(x, keyboard),
                    0x0015 => return self.opSetDelayTimer(x),
                    0x0018 => return self.opSetSoundTimer(x),
                    0x001E => return self.opAddToIndex(x),
                    0x0029 => return self.opFontCharacter(x),
                    0x0033 => return self.opBinaryCodedDecimalConversion(x),
                    0x0055 => return self.opStoreRegistersToMemory(x),
                    0x0065 => return self.opLoadRegistersFromMemory(x),
                    else => return 0,
                },
            else => return 0,
        }
    }

    // 0x00E0
    fn opClearScreen(display: *Display) u16 {
        display.clear();
        return 100;
    }

    // 0x1NNN
    fn opJump(self: *Self, nnn: u16) u16 {
        self.pc = nnn;
        return 105; 
    }

    // 0x6XNN
    fn opSetRegisterXToNN(self: *Self, x: u8, nn: u8) u16 {
        self.registers[x] = nn;
        return 27;
    }

    // 0x7XNN
    fn opAddNNToX(self: *Self, x: u8, nn: u8) u16 {
        var vx = self.registers[x];
        var vy = nn;
        while (vy != 0) {
            var carry = vx & vy;
            vx ^= vy;
            vy = carry << 1;
        }
        self.registers[x] = vx;
        return 45;
    }

    // 0xANNN
    fn opSetIndexToNNN(self: *Self, nnn: u16) u16 {
        self.index = nnn;
        return 55;
    }

    // 0xDXYN
    fn opDisplay(self: *Self, x: u8, y: u8, n: u8, display: *Display) u16 {
    	var index = self.index;
        var vx = @intCast(u16, self.registers[x]);
        var vy = @intCast(u16, self.registers[y]);
        std.debug.print("Rendering a {} pixel tall sprite at X: {}, Y: {} from the address: {}\n", .{n, vx, vy, index});

        var row = vy;
        while(row < vy + @intCast(u16,n)) : (row += 1) {
            if(row >= 32){
                break;
            }
            var startX = vx + row * 64;
            var endX = startX + 8;
            var bitMask: u8 = 0x80;
            var col = startX;
            while (col < endX) : (col += 1) {
                if ((col % 64) >= (startX % 64)) {
                    var position = (row * 64) + (col % 64);
                    if (self.memory.get(index) & bitMask == bitMask) {
                        display.flipPixel(position);
                    }
                }
                bitMask /= 2;
            }
            index += 1;
        }

        display.setDrawFlag(true);
        return 22734;
    }

    // 0xFX0A
    fn opGetKey(self: *Self, x: u16, keyboard: *Keyboard) u16 {
        var i: u8 = 0;
        while (i < 16) : (i += 1) {
            if (keyboard.hasKeyBeenReleased(i)) {
                self.registers[x] = i;
                return 1;
            }
        }
        self.pc -= 2;
        return 1;
    }

    // 0xFX29
    fn opFontCharacter(self: *Self, x: u16) u16 {
        self.index = SPRITE_CHARS_ADDR + self.registers[x];
        return 91;
    }

    // 0xFX55
    fn opStoreRegistersToMemory(self: *Self, x: u16) u16 {
        var i: u8 = 0;
        while (i <= x) : (i += 1) {
            self.memory.set(self.index+i, self.registers[i]);
        }
        return 605 + x * 64;
    }

    // 0xFX65
    fn opLoadRegistersFromMemory(self: *Self, x: u16) u16 {
        var i: u8 = 0;
        while (i <= x) : (i += 1) {
            self.registers[i] = self.memory.get(self.index + i);
        }
        return 605 + x * 64;
    }

    // 0xFX33
    fn opBinaryCodedDecimalConversion(self: *Self, x: u16) u16 {
        self.memory.set(self.index, self.registers[x] / 100);
        self.memory.set(self.index+1, (self.registers[x] / 10) % 10);
        self.memory.set(self.index+2, self.registers[x] % 10);
        return 927;
    }

    // 0xFX1E
    fn opAddToIndex(self: *Self, x: u16) u16 {
        self.index += self.registers[x];
        return 86;
    }

    // 0xFX07
    fn opGetDelayTimer(self: *Self, x: u16) u16 {
        self.registers[x] = self.delayTimer;
        return 45;
    }

    // 0xFX15
    fn opSetDelayTimer(self: *Self, x: u16) u16 {
        self.delayTimer = self.registers[x];
        return 45;
    }

    // 0xFX18
    fn opSetSoundTimer(self: *Self, x: u16) u16 {
        self.soundTimer = self.registers[x];
        return 45;
    }

    // 0x3XNN
    fn opSkipIfVxEqualsNn(self: *Self, x: u16, nn: u8) u16 {
        var clockCycles: u16 = 55;
        if (self.registers[x] == nn) {
            self.pc += 2;
        } else {
            clockCycles += 9;
        }
        return clockCycles;
    }

    // 0x4XNN
    fn opSkipIfVxNotEqualsNn(self: *Self, x: u16, nn: u8) u16 {
        var clockCycles: u16 = 55;
        if (self.registers[x] != nn) {
            self.pc += 2;
        } else {
            clockCycles += 9;
        }
        return clockCycles;
    }

    // 0x9XY0
    fn opSkipIfVxNotEqualsVy(self: *Self, x: u16, y: u16) u16 {
        var clockCycles: u16 = 73;
        if ((self).registers[x] != self.registers[y]) {
            self.pc += 2;
        } else {
            clockCycles += 9;
        }
        return clockCycles;
    }

    // 0x5XY0
    fn opSkipIfVxEqualsVy(self: *Self, x: u16, y: u16) u16 {
        var clockCycles: u16 = 55;
        if (self.registers[x] == self.registers[y]) {
            self.pc += 2;
        } else {
            clockCycles += 9;
        }
        return clockCycles;
    }

    // 0xEX9E
    fn opSkipIfKeyPressed(self: *Self, x: u16, keyboard: *Keyboard) u16 {
        if (keyboard.isKeyPressed(self.registers[x])) {
            self.pc += 2;
        }
        return 73;
    }

    // 0xEXA1
    fn opSkipIfNotKeyPressed(self: *Self, x: u16, keyboard: *Keyboard) u16 {
        if (!keyboard.isKeyPressed(self.registers[x])) {
            self.pc += 2;
        }
        return 73;
    }

    // 0xCXNN
    fn opRandom(self: *Self, x: u16, nn: u8) u16 {
        var random = rnd.random().int(u8);
        self.registers[x] = random & nn;
        return 164;
    }

    // 0xBXNN
    fn opJumpWithOffset(self: *Self, addr: u16) u16 {
        self.pc = addr + self.registers[0];
        return 105;
    }

    // 0x8XY0
    fn opSetVxToValueOfVy(self: *Self, x: u16, y: u16) u16 {
        self.registers[x] = self.registers[y];
        return 200;
    }

    // 0x8XY6
    fn opShiftRight(self: *Self, x: u16) u16 {
        var flag = self.registers[x] & 0x1;
        self.registers[x] >>= 1;
        self.registers[0xF] = flag;
        return 200;
    }

    // 0x8XYE
    fn opShiftLeft(self: *Self, x: u16) u16 {
        var flag = (self.registers[x] & 0x80) >> 7;
        self.registers[x] <<= 1;
        self.registers[0xF] = flag;
        return 200;
    }

    // 0x8XY5
    fn opSubtractYFromX(self: *Self, x: u16, y: u16) u16 {
        var vx = self.registers[x];
        var vy = self.registers[y];
        var borrow: u8 = 0;
        if (vx > vy) {
            borrow = 1;
        }
        self.registers[x] = vx -% vy;
        self.registers[0xF] = borrow;
        return 200;
    }

    // 0x8XY7
    fn opSubtractXFromY(self: *Self, x: u16, y: u16) u16 {
        var vx = self.registers[x];
        var vy = self.registers[y];
        var flag: u8 = 0;
        if (vy > vx) {
            flag = 1;
        }
        self.registers[x] = vy -% vx;
        self.registers[0xF] = flag;
        return 200;
    }

    // 0x8XY4
    fn opAddWithCarry(self: *Self, x: u8, y: u8) u16 {
        var sum = @intCast(u16, self.registers[x]) + @intCast(u16, self.registers[y]);
        if (sum > 255) {
            self.registers[0xF] = 1;
        } else {
            self.registers[0xF] = 0;
        }
        self.registers[x] = @intCast(u8, sum & 0xFF);
        return 200;
    }

    // 0x8XY1
    fn opBinaryOr(self: *Self, x: u16, y: u16) u16 {
        self.registers[x] |= self.registers[y];
        return 200;
    }

    // 0x8XY3
    fn opBinaryXor(self: *Self, x: u16, y: u16) u16 {
        self.registers[x] ^= self.registers[y];
        return 200;
    }

    // 0x8XY2
    fn opBinaryAnd(self: *Self, x: u16, y: u16) u16 {
        self.registers[x] &= self.registers[y];
        return 200;
    }

    // 0x000E
    fn opReturn(self: *Self) u16 {
        self.pc = self.index;
        return 1;
    }

    // 0x2NNN
    fn opJumpToSubroutine(self: *Self, addr: u16) u16 {
        if (self.sp < self.stack.len) {
            self.stack[self.sp] = self.pc;
            self.sp += 1;
            self.pc = addr;
        }
        return 105;
    }

    // 0x00EE
    fn opReturnFromSubroutine(self: *Self) u16 {
        if (self.sp > 0) {
            self.sp -= 1;
            self.pc = self.stack[self.sp];
        }
        return 105;
    }
};

File: typescript/src/cpu.ts
/// <reference path="registers.ts"/>
/// <reference path="memory.ts"/>
/// <reference path="keyboard.ts"/>
/// <reference path="display.ts"/>

const FRAME_TICKS = 16666;

class CPU {
    _registers: Registers;
    _memory : IMemory;
    _keyboard : IKeyboard;
    _display : IDisplay;
    _audio : IAudio;
    _instructionsCounter : number;
    _pc : number;
    _drawFlag : boolean;
    _stack : number[];
    _i : number;
    _microseconds : number;

    public DelayTimer : number;
    public SoundTimer : number;
    
    get PC() : number { return this._pc; }
    get I() : number { return this._i; }

    constructor(keyboard : IKeyboard, display : IDisplay, audio : IAudio) {
        this._keyboard = keyboard;
        this._display = display;
        this._audio = audio;
    }

    attachMemory(memory : IMemory) : void {
        this._display.clear();
        this._audio.stop();
        this._registers = new Registers();
        this._pc = 0x200;
        this._instructionsCounter = 0;
        this._stack = [];
        this._microseconds = 0;
        this._memory = memory;
    }

    tick() : void {
        if(this._memory == null)
            return;
        
        if (this.DelayTimer > 0) {
            this.DelayTimer -= 1;
        }
        if (this.SoundTimer> 0) {
            this.SoundTimer -= 1;
            if (this._audio.isPlaying() == false) {
                this._audio.start();
            }
        } else if (this._audio.isPlaying()) {
            this._audio.stop();
        }
        this._microseconds += FRAME_TICKS;
        while(1) {
            let delta = this.emulateCPUCycle();
            this._microseconds -= delta;
            if (this._microseconds <= 0 || delta == 0) {
                break;
            }
        }
    }

    emulateCPUCycle() : number {
        this._instructionsCounter++;
        let opcode = this._memory.getOpCode(this._pc);

        let x = (opcode & 0x0F00) >> 8;
        let y = (opcode & 0x00F0) >> 4;
        let vx = this._registers.getRegister(x);
        let vy = this._registers.getRegister(y);
        let n = opcode & 0x000F;
        let nn = opcode & 0x00FF;
        let nnn = opcode & 0x0FFF;

        this._pc += 2;

        switch (opcode)
        {
            case 0x00E0: // Clears the screen
                this._display.clear();
                this._drawFlag = true;
                return 109;

            case 0x00EE: // Returns from a subroutine.
                this._pc = this._stack.pop();
                return 105;
            
            case 0x000E:
                this._pc = this._i;
                return 1;

            default:
                break;
        }

        switch (opcode & 0xF000)
        {
            case 0x1000: // Jumps to address nnn.
                this._pc = nnn;
                return 105;

            case 0x2000: // Calls subroutine at nnn.
                this._stack.push(this._pc);
                this._pc = nnn;
                return 105;

            case 0x3000: {// Skips the next instruction if vx equals nn. (Usually the next instruction is a jump to skip a code block)
                let clockCycles = 55;    
                if (vx == nn) {
                    this._pc += 2;
                } else {
                    clockCycles += 9;
                }
                return clockCycles;
            }

            case 0x4000: {// Skips the next instruction if vx doesn't equal nn. (Usually the next instruction is a jump to skip a code block)
                let clockCycles = 55;
                if (vx != nn) {
                    this._pc += 2;
                } else {
                    clockCycles += 9;
                }
                return clockCycles;
            }
            case 0x5000: {// Skips the next instruction if vx equals vy. (Usually the next instruction is a jump to skip a code block)
                let clockCycles = 55;
                if (vx == vy) {
                    this._pc += 2;
                } else {
                    clockCycles += 9;
                }
                return clockCycles;
            }

            case 0x6000: // Sets vx to nn.
                this._registers.setRegister(x, nn);
                return 27;

            case 0x7000: // Adds nn to vx. (Carry flag is not changed)
                this._registers.apply(x, vx => (vx + nn) & 0xFF);
                return 45;

            case 0x8000: {
                switch (n) {
                    case 0x0000: // Sets vx to the value of vy.
                        this._registers.setRegister(x, vy);
                        return 200;

                    case 0x0001: // Sets vx to vx or vy. (Bitwise OR operation)
                        this._registers.apply(x, vx => vx | vy);
                        return 200;

                    case 0x0002: // Sets vx to vx and vy. (Bitwise AND operation)
                        this._registers.apply(x, vx => vx & vy);
                        return 200;

                    case 0x0003: // Sets vx to vx xor vy.
                        this._registers.apply(x, vx => vx ^ vy);
                        return 200;

                    case 0x0004: // Adds vy to vx. VF is set to 1 when there's a carry, and to 0 when there isn't.
                        this._registers.apply(x, vx => {
                            var sum = vx + vy;
                            this._registers.setRegister(0xF, sum > 0xFF ? 1 : 0);
                            return sum & 0xFF;
                        });
                        return 200;

                    case 0x0005: // vy is subtracted from vx. VF is set to 0 when there's a borrow, and 1 when there isn't.
                        this._registers.setRegister(0xF, vy > vx ? 0 : 1);
                        this._registers.apply(x, vx => vx - vy);
                        return 200;

                    case 0x0006: // Stores the least significant bit of vx in VF and then shifts vx to the right by 1.[2]
                        this._registers.setRegister(0xF, vx & 0x1);
                        this._registers.apply(x, vx => vx >> 1);
                        return 200;

                    case 0x0007: // Sets vx to vy minus vx. VF is set to 0 when there's a borrow, and 1 when there isn't.
                        this._registers.setRegister(0xF, vx > vy ? 0 : 1);
                        this._registers.apply(x, vx => (vy - vx) & 0xFF);
                        return 200;

                    case 0x000E: // Stores the most significant bit of vx in VF and then shifts vx to the left by 1.[3]
                        this._registers.setRegister(0xF, (vx & 0x80) > 0 ? 1 : 0);
                        this._registers.apply(x, vx => (vx << 1) & 0xFF);
                        return 200;
                    default:
                        break;
                    }
                }
                break;

            case 0x9000: {// Skips the next instruction if vx doesn't equal vy. (Usually the next instruction is a jump to skip a code block)
                let clockCycles = 73;
                if (vx != vy) {
                    this._pc += 2;
                } else {
                    clockCycles += 9;
                }
                return clockCycles;
            }
            case 0xA000: // Sets I to the address nnn.
                this._i = nnn;
                return 55;

            case 0xB000: // Jumps to the address nnn plus V0..
                this._pc = (this._registers.getRegister(0) + nnn) & 0x0FFF;
                return 105;

            case 0xC000: // Sets vx to the result of a bitwise and operation on a random number (Typically: 0 to 255) and nn.
                // TODO: this._registers.setRegister(x, this._random.NextByte() & nn);
                return 164;

            case 0xD000: // Draws a sprite at coordinate (vx, vy) that has a width of 8 pixels and a height of n pixels. Each row of 8 pixels is read as bit-coded starting from memory location I; I value doesn’t change after the execution of this instruction. As described above, VF is set to 1 if any screen pixels are flipped from set to unset when the sprite is drawn, and to 0 if that doesn’t happen
                let height = n;
                this._registers.setRegister(0xF, 0);

                for (let yLine = 0; yLine < height; yLine++) {
                    let pixelValue = this._memory.getByte(this._i + yLine);

                    for (let xLine = 0; xLine < 8; xLine++) {
                        if (CPU.getBitValue(pixelValue, xLine) != 0) {
                            let xCoord = vx + xLine;
                            let yCoord = vy + yLine;

                            if (xCoord >= this._display.Width) {
                                xCoord %= this._display.Width;
                            }

                            if (yCoord >= this._display.Height) {
                                yCoord %= this._display.Height;
                            }

                            if (this._display.getPixel(xCoord, yCoord)) {
                                this._registers.setRegister(0xF, 1);
                            }

                            this._display.setPixel(xCoord, yCoord);
                        }
                    }
                }

                this._display.setDrawFlag();
                return 22734;

            case 0xE000:
                switch (nn) {
                    case 0x9E: // Skips the next instruction if the key stored in vx is pressed. (Usually the next instruction is a jump to skip a code block)
                        this._pc += this._keyboard.isKeyPressed(vx) ? 2 : 0;
                        return 73;

                    case 0xA1: // Skips the next instruction if the key stored in vx isn't pressed. (Usually the next instruction is a jump to skip a code block)
                        this._pc += this._keyboard.isKeyPressed(vx) ? 0 : 2;
                        return 73;

                    default:
                        break;
                }
                break;

            case 0xF000:

                switch (nn) {
                    case 0x07: // Sets vx to the value of the delay timer.
                        this._registers.setRegister(x, this.DelayTimer & 0xFF);
                        return 45;

                    case 0x0A: // A key press is awaited, and then stored in vx. (Blocking Operation. All instruction halted until next key event)
                        for (let i = 0; i < this._keyboard.getAllKeys().length; i++) {
                            if (this._keyboard.isKeyPressed(i)) {
                                this._registers.setRegister(x, i);
                                break;
                            }
                        }
                        this._pc -= 2;
                        return 1;

                    case 0x15: // Sets the delay timer to vx.
                        this.DelayTimer = vx & 0xFF;
                        return 45;

                        case 0x18: // Sets the sound timer to vx.
                        this.SoundTimer = vx & 0xFF;
                        return 45;

                    case 0x1E: // Adds vx to I.
                        this._i += vx & 0xFF;
                        return 86;

                    case 0x29: // Sets I to the location of the sprite for the character in vx. Characters 0-F (in hexadecimal) are represented by a 4x5 font.
                        this._i = 0x50 + (vx * 5);
                        return 91;

                    case 0x33: // vx, with the most significant of three digits at the address in I, the middle digit at I plus 1, and the least significant digit at I plus 2. (In other words, take the decimal representation of vx, place the hundreds digit in memory at location in I, the tens digit at location I+1, and the ones digit at location I+2.)
                        this._memory.setByte(this._i, vx / 100);
                        this._memory.setByte(this._i + 1, (vx % 100) / 10);
                        this._memory.setByte(this._i + 2, vx % 10);
                        return 927;

                    case 0x55: // Stores V0 to vx (including vx) in memory starting at address I. The offset from I is increased by 1 for each value written, but I itself is left unmodified.
                        for (let p = 0; p <= x; p++) {
                            this._memory.setByte(this._i + p, this._registers.getRegister(p));
                        }
                        return 605 + x * 64;

                    case 0x65: // Fills V0 to vx (including vx) with values from memory starting at address I. The offset from I is increased by 1 for each value written, but I itself is left unmodified.
                        for (let p = 0; p <= x; p++) {
                            this._registers.setRegister(p, this._memory.getByte(this._i + p) & 0xFF);
                        }
                        return 605 + x * 64;

                    default:
                        break;
                    }

                break;

            default:
                break;
        }
        return 0;
    }

    private static getBitValue(value : number, bitIndex : number) : number {
        return value & (0x80 >> bitIndex);
    }
}
